erpnext/accounts/doctype/bulk_statement_of_accounts/bulk_statement_of_accounts.py

```diff
@@ -0,0 +1,241 @@
+from frappe.core.doctype.communication.email import make
+
+from frappe.utils.print_format import report_to_pdf
+from frappe.utils.pdf import get_pdf
+from frappe.utils import today, add_days, add_months, getdate, format_date
+from datetime import timedelta
+from frappe.www.printview import get_print_style
+
+to_add = { 'Monthly': 1, 'Quarterly': 3 }
+
+class BulkStatementOfAccounts(Document):
+	def validate(self):
+		if not self.customer_list:
+			frappe.throw(frappe._('Customers not selected!'))
+
+		global to_add
+		if self.enable_auto_email:
+			if self.frequency == 'Weekly':
+				self.to_date = add_days(self.start_date, to_add[self.frequency])
+			else:
+				self.to_date = add_months(self.start_date, to_add[self.frequency])
+			self.from_date = add_months(self.to_date, -1 * self.filter_duration)
+
+def get_report_pdf(doc, consolidated=True):
+	statement_dict = {}
+	aging = ''
+	base_template_path = "frappe/www/printview.html"
+
+	for entry in doc.customer_list:
+		if doc.include_ageing:
+			ageing_filters = frappe._dict({
+				'company': doc.company,
+				'report_date': doc.to_date,
+				'ageing_based_on': doc.ageing_based_on,
+				'range1': 30,
+				'range2': 60,
+				'range3': 90,
+				'range4': 120,
+				'customer': entry.customer
+			})
+			col1, aging = get_ageing(ageing_filters)
+			aging[0]['ageing_based_on'] = doc.ageing_based_on
+
+		tax_id = frappe.get_doc('Customer', entry.customer).tax_id
+
+		filters= frappe._dict({
+			'from_date': doc.from_date,
+			'to_date': doc.to_date,
+			'company': doc.company,
+			'finance_book': doc.finance_book if doc.finance_book else '',
+			'account': doc.account if doc.account else '',
+			'party_type': 'Customer',
+			'party': [entry.customer],
+			'group_by': doc.group_by,
+			'currency': doc.currency,
+			'cost_center': [cc.cost_center_name for cc in doc.cost_center],
+			'project': [p.project_name for p in doc.project],
+			'show_opening_entries': 0,
+			'include_default_book_entries': 0,
+			'show_cancelled_entries': 1,
+			'tax_id': tax_id if tax_id else ''
+		})
+		col, res = get_soa(filters)
+		if len(res) == 3:
+			continue
+		html = frappe.render_template('accounts/doctype/bulk_statement_of_accounts/bsoa.html', \
+			{"filters": filters, "data": res, "aging": aging[0] if doc.include_ageing else None})
+		html = frappe.render_template(base_template_path, {"body": html, "css": get_print_style(), "title": "Statement For " + entry.customer})
+		statement_dict[entry.customer] = html
+	if not bool(statement_dict):
+		return False
+	elif consolidated:
+		result = ''.join(list(statement_dict.values()))
+		return get_pdf(result, {'orientation': doc.orientation})
+	else:
+		for customer, statement_html in statement_dict.items():
+			statement_dict[customer]=get_pdf(statement_html, {'orientation': doc.orientation})
+		return statement_dict
+
+@frappe.whitelist()
+def fetch_customers(customer_collection, collection_name, primary_mandatory):
```

Comment:
split into smaller functions. Each customer_collection can be a separate function.